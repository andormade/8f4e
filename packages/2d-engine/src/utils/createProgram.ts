export default function createProgram(
	gl: WebGL2RenderingContext | WebGLRenderingContext,
	shaders: WebGLShader[]
): WebGLProgram {
	const program = gl.createProgram();

	shaders.forEach(shader => {
		gl.attachShader(program, shader);
	});

	gl.linkProgram(program);

	const linked = gl.getProgramParameter(program, gl.LINK_STATUS);

	if (!linked) {
		const lastError = gl.getProgramInfoLog(program);
		console.error('Error in program linking: ' + lastError);
		gl.deleteProgram(program);
		return null;
	}

	return program;
}
