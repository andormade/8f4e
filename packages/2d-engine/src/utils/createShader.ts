export default function createShader(
	gl: WebGL2RenderingContext | WebGLRenderingContext,
	shaderSource: string,
	shaderType: number
): WebGLShader {
	const shader = gl.createShader(shaderType);
	gl.shaderSource(shader, shaderSource);
	gl.compileShader(shader);

	if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
		const info = gl.getShaderInfoLog(shader);
		console.error('Could not compile web gl shader', info);
	}

	return shader;
}
