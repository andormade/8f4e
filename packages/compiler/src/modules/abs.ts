import { ModuleGenerator } from '../types';
import { compile } from 'module-compiler';

const abs: ModuleGenerator<unknown> = function (moduleId, offset) {
	return compile(
		`
		# abs module
		
		# memory
		private DEFAULT_VALUE 0
		inputPointer in DEFAULT_VALUE
		output out 0
		
		# locals
		local input
		
		# code
		const out
		const in
		load
		load
		localSet input
		localGet input
		const 0
		lessThan
		if 
			const 0
			localGet input
			sub
		else
			localGet input
		end
		store
	`,
		moduleId,
		offset.byte(0)
	);
};

export default abs;
