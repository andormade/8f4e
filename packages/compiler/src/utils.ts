import { GLOBAL_ALIGNMENT_BOUNDARY } from './consts';
import { BlockStack, CompiledModule, MemoryMap, StackItem } from './types';

export function isMemoryIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return memoryMap.has(name);
}

export function isMemoryReferenceIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return (
		(name.startsWith('&') && memoryMap.has(name.substring(1))) ||
		(name.endsWith('&') && memoryMap.has(name.slice(0, -1)))
	);
}

export function isMemoryPointerIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('*') && memoryMap.has(name.substring(1));
}

export function isWordSpanIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('$') && memoryMap.has(name.substring(1));
}

export function isWordSizeIdentifier(memoryMap: MemoryMap, name: string): boolean {
	return name.startsWith('%') && memoryMap.has(name.substring(1));
}

export function getDataStructure(memoryMap: MemoryMap, id: string) {
	return memoryMap.get(id);
}

export function getDataStructureByteAddress(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.byteAddress : 0;
}

export function getMemoryStringLastAddress(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.byteAddress + (memoryItem.wordAlignedSize - 1) * GLOBAL_ALIGNMENT_BOUNDARY : 0;
}

export function getWordSpan(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.wordAlignedSize : 0;
}

export function getWordSize(memoryMap: MemoryMap, id: string): number {
	const memoryItem = getDataStructure(memoryMap, id);
	return memoryItem ? memoryItem.elementWordSize : 0;
}

export function isInstructionIsInsideAModule(blockStack: BlockStack) {
	for (let i = blockStack.length - 1; i > 0; i--) {
		if (blockStack[i].isModuleBlock) {
			return true;
		}
	}
	return false;
}

export function isInstructionIsInsideAGroup(blockStack: BlockStack) {
	return blockStack[blockStack.length - 1] && blockStack[blockStack.length - 1].isGroupBlock;
}

export function calculateMemoryWordSize(memory: MemoryMap): number {
	return Array.from(memory.values()).reduce((accumulator, current) => {
		return accumulator + current.wordAlignedSize;
	}, 0);
}

export function areAllOperandsIntegers(...operands: StackItem[]): boolean {
	return !operands.some(operand => !operand.isInteger);
}

export function areAllOperandsFloats(...operands: StackItem[]): boolean {
	return !operands.some(operand => operand.isInteger);
}
