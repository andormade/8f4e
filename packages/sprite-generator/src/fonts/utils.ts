export function pad(paddingLeft: number, arr: number[], paddingRight = 0): number[] {
	return [...new Array(paddingLeft).fill(0b00000000), ...arr, ...new Array(paddingRight).fill(0b00000000)];
}

export function same(byte: number | number[], times: number): number[] {
	return new Array(times).fill(byte).flat();
}

export function mirr(toBeMirrored: number[], middle: number[] = []): number[] {
	return [...toBeMirrored, ...middle, ...toBeMirrored.slice().reverse()];
}
