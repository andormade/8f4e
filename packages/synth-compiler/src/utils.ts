import { RelativeAddressCalculator } from './types';

export function createRelativeAddressCalculator(offset: number, wordLength: number): RelativeAddressCalculator {
	return {
		byte: function (nthWord) {
			return nthWord * wordLength + offset * wordLength;
		},
		word: function (nthWord) {
			return nthWord + offset;
		},
	};
}
