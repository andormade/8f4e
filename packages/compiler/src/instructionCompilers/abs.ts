import { ErrorCode, getError } from '../errors';
import { InstructionCompiler } from '../types';
import WASMInstruction from '../wasmUtils/wasmInstruction';
import { isInstructionIsInsideAModule, saveByteCode } from '../utils';
import { compileSegment } from '../compiler';

const abs: InstructionCompiler = function (line, context) {
	if (!isInstructionIsInsideAModule(context.blockStack)) {
		throw getError(ErrorCode.INSTRUCTION_INVALID_OUTSIDE_BLOCK, line, context);
	}

	const operand = context.stack.pop();

	if (!operand) {
		throw getError(ErrorCode.INSUFFICIENT_OPERANDS, line, context);
	}

	if (operand.isInteger) {
		// For some reason there is still no abs instruction for integers in Web Assembly.
		context.stack.push({ isInteger: true, isNonZero: operand.isNonZero });
		const valueName = '__absify_value' + line.lineNumber;

		return compileSegment(
			[
				`local int ${valueName}`,
				`localSet ${valueName}`,
				`localGet ${valueName}`,
				'push 0',
				'lessThan',
				'if',
				' push 0',
				` localGet ${valueName}`,
				' sub',
				'else',
				` localGet ${valueName}`,
				'ifEnd',
			],
			context
		);
	} else {
		context.stack.push({ isInteger: false, isNonZero: operand.isNonZero });
		return saveByteCode(context, [WASMInstruction.F32_ABS]);
	}
};

export default abs;
