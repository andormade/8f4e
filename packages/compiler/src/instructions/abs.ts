import { InstructionHandler } from '../types';
import { ErrorCode, getError } from '../errors';
import { parseSegment } from '../compiler';
import { isInstructionIsInsideAModule } from '../utils';
import WASMInstruction from '../wasmUtils/wasmInstruction';

const cycle: InstructionHandler = function (line, context) {
	if (isInstructionIsInsideAModule(context.blockStack)) {
		throw getError(ErrorCode.INSTRUCTION_INVALID_OUTSIDE_BLOCK, line, context);
	}

	const operand = context.stack.pop();

	if (!operand) {
		throw getError(ErrorCode.INSUFFICIENT_OPERANDS, line, context);
	}

	if (operand.isInteger) {
		// For some reason there is still no abs instruction for integers in Web Assembly.
		context.stack.push({ isInteger: true });
		const valueName = '__absify_value' + line.lineNumber;

		return parseSegment(
			[
				`local int ${valueName}`,
				`localSet ${valueName}`,
				`localGet ${valueName}`,
				'push 0',
				'lessThan',
				'if',
				' push 0',
				` localGet ${valueName}`,
				' sub',
				'else',
				` localGet ${valueName}`,
				'end',
			],
			context
		);
	} else {
		context.stack.push({ isInteger: false });
		return {
			byteCode: [WASMInstruction.F32_ABS],
			context,
		};
	}
};

export default cycle;
