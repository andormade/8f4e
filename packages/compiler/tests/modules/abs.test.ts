import { createTestModule } from '../utils';
import abs, { Memory } from '../../src/modules/abs';

let testModule;

test('if compiled module matches with snapshot', () => {
	expect(abs('id', () => 0)).toMatchSnapshot();
});

describe('functional tests', () => {
	beforeAll(async () => {
		testModule = await createTestModule(abs);
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('abs module', () => {
		const { memory, test } = testModule;

		memory[Memory.DEFAULT_VALUE] = 1;
		test();
		expect(memory[Memory.OUTPUT]).toBe(1);

		memory[Memory.DEFAULT_VALUE] = -1;
		test();
		expect(memory[Memory.OUTPUT]).toBe(1);

		memory[Memory.DEFAULT_VALUE] = -69;
		test();
		expect(memory[Memory.OUTPUT]).toBe(69);

		memory[Memory.DEFAULT_VALUE] = 420;
		test();
		expect(memory[Memory.OUTPUT]).toBe(420);
	});
});
