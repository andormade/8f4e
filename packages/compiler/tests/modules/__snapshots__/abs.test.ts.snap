// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if compiled module matches with snapshot 1`] = `
Object {
  "functionBody": Array [
    35,
    1,
    1,
    127,
    65,
    0,
    65,
    0,
    40,
    2,
    0,
    40,
    2,
    0,
    33,
    0,
    32,
    0,
    65,
    0,
    72,
    4,
    127,
    65,
    0,
    32,
    0,
    107,
    5,
    32,
    0,
    11,
    54,
    2,
    0,
    11,
  ],
  "initialMemory": Array [
    0,
    0,
    0,
    0,
  ],
  "memoryAddresses": Array [
    Object {
      "address": 0,
      "id": "out",
    },
    Object {
      "address": 0,
      "default": 0,
      "id": "in",
    },
  ],
  "moduleId": "id",
  "offset": 0,
}
`;